### PR TITLE
feat: add manage:list-backends command with cleanup verdicts

### PR DESCRIPTION
**Pull request by jirkasemmler**
Changes:

- New `manage:list-backends` command — read-only inventory of a stack's storage backends with a cleanup verdict per backend (`DELETE_UNSUPPORTED_UNUSED` / `DELETE_UNUSED` / `REVIEW_UNSUPPORTED_IN_USE` / `REVIEW_MAINTAINER_ONLY` / `KEEP`). Prints comma-separated ID lists ready to paste into `manage:delete-backend`.
- Fetches the detail of every backend (`GET /manage/storage-backend/{id}`) to add columns missing from the list response: `loginType`, `keyRotated`, `dynBackends`, `useSso`, `ssoEnabled`, `ssoConfigured`.
- `--format=table|csv` (CSV keeps stdout parseable, summary goes to stderr), `--verdict=X` and `--unsupported` filters; summary is always computed over the whole stack.
- Registered in `cli.php`, documented in README.

Used for the [DMD-1638](https://linear.app/keboola/issue/DMD-1638/cleanup-all-prod-backends) prod backend cleanup — drove the inventory and keypair review of all 14 prod stacks.

---

Additional notes

:warning:  *Don't forget to release new version after merge*

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/Keboola/Console/Command/ListStorageBackends.php [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+    }
+
+    /**
+     * @param array<mixed> $backend
+     * @return array<mixed>
+     */
+    private function fetchDetail(GuzzleClient $guzzle, array $backend): array
+    {
+        try {
+            $response = $guzzle->get(sprintf('manage/storage-backend/%d', $this->asInt($backend['id'] ?? 0)));
```
Comment (romanbracinik) [inline src/Keboola/Console/Command/ListStorageBackends.php]: Preco je tu ten guzzle ? Manage api klient ma implementovany detail storage https://github.com/keboola/kbc-manage-api-php-client/blob/master/src/Client.php#L1198

File: src/Keboola/Console/Command/ListStorageBackends.php [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+
+        $visibleRows = $this->filterRows($rows, $unsupportedOnly, $verdictFilter);
+
+        if ($format === self::FORMAT_CSV) {
+            $this->renderCsv($output, $visibleRows);
+        } else {
+            $this->renderTable($output, $visibleRows);
+        }
+
+        $this->renderSummary($output, $rows, $format === self::FORMAT_CSV);
```
Comment (romanbracinik) [inline src/Keboola/Console/Command/ListStorageBackends.php]: **Bug:** `renderSummary()` receives the unfiltered `$rows`, so with `--verdict`/`--unsupported` the summary counts and the paste-ready "Delete candidates" ID lists ignore the filter — the table above shows filtered rows, but the ID list still covers all backends. Since this output is meant to be pasted into `manage:delete-backend`, an operator can end up deleting backends outside the scope they filtered for. `Total backends: N` also contradicts the number of rows shown. Pass `$visibleRows` to `renderSummary()` (or apply the same filter inside).

File: src/Keboola/Console/Command/ListStorageBackends.php [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+        assert(is_string($token));
+        $url = $input->getArgument(self::ARGUMENT_URL);
+        assert(is_string($url));
+
+        $format = $input->getOption(self::OPTION_FORMAT);
+        if (!in_array($format, [self::FORMAT_TABLE, self::FORMAT_CSV], true)) {
+            $output->writeln('<error>Unknown format, use table or csv</error>');
+            return 1;
+        }
+        $verdictFilter = $input->getOption(self::OPTION_VERDICT);
```
Comment (romanbracinik) [inline src/Keboola/Console/Command/ListStorageBackends.php:97]: **Bug:** `--verdict` is never validated against the `VERDICT_*` constants, while `--format` a few lines above is. `filterRows()` compares with strict equality, so a typo or lowercase value (`--verdict=delete_unused`) silently matches nothing and prints an empty table with exit 0 — easy to read as "no backends with this verdict". Validate the value and fail with the list of allowed verdicts.

File: src/Keboola/Console/Command/ListStorageBackends.php [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+     */
+    private function renderCsv(OutputInterface $output, array $rows): void
+    {
+        $output->writeln(implode(',', self::COLUMNS));
+        foreach ($rows as $row) {
+            $cells = [];
+            foreach (self::COLUMNS as $column) {
+                $cells[] = '"' . str_replace('"', '""', (string) $row[$column]) . '"';
+            }
+            $output->writeln(implode(',', $cells));
```
Comment (romanbracinik) [inline src/Keboola/Console/Command/ListStorageBackends.php]: CSV rows are written via `$output->writeln()` without `OutputInterface::OUTPUT_RAW`, so the Symfony output formatter interprets tag-like content in cell values (e.g. an owner/host containing `<info>` or any `<...>` matching a style) — the CSV on stdout can contain ANSI escapes or silently dropped text. Use `$output->writeln($line, OutputInterface::OUTPUT_RAW)` for CSV output.

File: src/Keboola/Console/Command/ListStorageBackends.php (hunk 1 above)
Comment (romanbracinik) [inline src/Keboola/Console/Command/ListStorageBackends.php]: A backend with a missing or non-numeric `id` is coerced to 0 by `asInt()`, producing a bogus `GET manage/storage-backend/0` (404 swallowed) and a failed-details warning naming backend "0" — an id that doesn't exist in the table. The 0 is also used as the row's sort key and rendered in the id column. Consider skipping (or failing loudly on) entries without a valid id.

File: src/Keboola/Console/Command/ListStorageBackends.php [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+            $output->writeln('<error>Unknown format, use table or csv</error>');
+            return 1;
+        }
+        $verdictFilter = $input->getOption(self::OPTION_VERDICT);
+        assert($verdictFilter === null || is_string($verdictFilter));
+        $unsupportedOnly = (bool) $input->getOption(self::OPTION_UNSUPPORTED);
+
+        $client = new Client(['url' => $url, 'token' => $token]);
+        $guzzle = new GuzzleClient([
+            'base_uri' => $url,
```
Comment (romanbracinik) [inline src/Keboola/Console/Command/ListStorageBackends.php]: `GuzzleHttp\Client` is used directly, but `guzzlehttp/guzzle` is not declared in `composer.json` `require` — it's only a transitive dependency of the Keboola SDK clients. If an SDK ever switches HTTP client, this command dies at runtime with "Class not found". Add it to `require` explicitly.

File: src/Keboola/Console/Command/ListStorageBackends.php [hunk 6]
```diff
@@ -0,0 +1,364 @@
...
+     * @param array<mixed> $backend
+     * @return array<mixed>
+     */
+    private function fetchDetail(GuzzleClient $guzzle, array $backend): array
+    {
+        try {
+            $response = $guzzle->get(sprintf('manage/storage-backend/%d', $this->asInt($backend['id'] ?? 0)));
+            $detail = json_decode((string) $response->getBody(), true);
+            return is_array($detail) ? $detail : [];
+        } catch (Throwable) {
```
Comment (romanbracinik) [inline src/Keboola/Console/Command/ListStorageBackends.php]: The raw Guzzle detail calls have no retry, unlike the ManageApi `Client`'s built-in backoff, and `catch (Throwable)` swallows everything — a single transient 429/500 blanks that row's detail columns and forces a full re-run. Consider Guzzle retry middleware, or better, upstreaming a `getStorageBackend()` method into the manage-api SDK so this workaround isn't needed.

File: src/Keboola/Console/Command/ListStorageBackends.php [hunk 7]
```diff
@@ -0,0 +1,364 @@
...
+        if ($unsupportedOnly) {
+            $rows = array_filter($rows, fn (array $r) => !in_array($r['backend'], self::KEPT_BACKENDS, true));
+        }
+        if ($verdict !== null) {
+            $rows = array_filter($rows, fn (array $r) => $r['verdict'] === $verdict);
+        }
+        return array_values($rows);
+    }
+
+    private function resolveVerdict(string $type, int $projects, int $maintainers): string
```
Comment (romanbracinik) [inline src/Keboola/Console/Command/ListStorageBackends.php]: `resolveVerdict()`/`buildRow()`/`filterRows()` are pure logic whose output (the `DELETE_*` ID lists) is documented as ready to paste into the destructive `manage:delete-backend`, yet the PR adds no tests. Per the repo pattern (AGENTS.md, `DataAppOrchestratorTaskMigrator`), extract this into a plain class and unit-test it — a future regression in the verdict rules would otherwise pass CI silently.

File: src/Keboola/Console/Command/ListStorageBackends.php [hunk 8]
```diff
@@ -0,0 +1,364 @@
...
+        assert(is_array($backends));
+
+        $progress = $output instanceof ConsoleOutputInterface ? $output->getErrorOutput() : $output;
+        $progress->writeln(sprintf('Fetching details for %d backends...', count($backends)));
+
+        $rows = [];
+        $failedDetails = [];
+        foreach ($backends as $backend) {
+            assert(is_array($backend));
+            $detail = $this->fetchDetail($guzzle, $backend);
```
Comment (romanbracinik) [inline src/Keboola/Console/Command/ListStorageBackends.php]: Detail is fetched for every backend before filtering, but `resolveVerdict()` and both filters only need list-response fields, and the detail-only columns are shown only for visible rows. With `--verdict=...` on a large stack this issues one needless GET per hidden backend. Compute verdicts and filter first, then fetch detail only for visible rows.

File: src/Keboola/Console/Command/ListStorageBackends.php [hunk 9]
```diff
@@ -0,0 +1,364 @@
...
+
+        $backends = $client->listStorageBackend();
+        assert(is_array($backends));
+
+        $progress = $output instanceof ConsoleOutputInterface ? $output->getErrorOutput() : $output;
+        $progress->writeln(sprintf('Fetching details for %d backends...', count($backends)));
+
+        $rows = [];
+        $failedDetails = [];
+        foreach ($backends as $backend) {
```
Comment (romanbracinik) [inline src/Keboola/Console/Command/ListStorageBackends.php:118]: The per-backend detail GETs run strictly sequentially — one blocking round-trip each, with a 30s timeout burned per failing backend. On a stack with many backends this takes minutes. `GuzzleHttp\Pool` (or `Promise\Utils::settle`) with a small concurrency limit would do the same work in a fraction of the wall time.

File: src/Keboola/Console/Command/ListStorageBackends.php [hunk 10]
```diff
@@ -0,0 +1,364 @@
...
+        return $value ? '1' : '0';
+    }
+
+    /**
+     * @param array<int, array<string, int|string>> $rows
+     */
+    private function renderCsv(OutputInterface $output, array $rows): void
+    {
+        $output->writeln(implode(',', self::COLUMNS));
+        foreach ($rows as $row) {
```
Comment (romanbracinik) [inline src/Keboola/Console/Command/ListStorageBackends.php:209]: This hand-rolls the CSV quoting/escaping that `fputcsv()` already provides (RFC 4180 edge cases included). Writing rows with `fputcsv()` into a `php://temp` stream and passing the result to `$output` removes the bespoke escaping code.

**Review by romanbracinik — CHANGES_REQUESTED** [review]

File: src/Keboola/Console/Command/ListStorageBackends.php [hunk 11]
```diff
@@ -0,0 +1,364 @@
...
+use Symfony\Component\Console\Input\InputOption;
+use Symfony\Component\Console\Output\ConsoleOutputInterface;
+use Symfony\Component\Console\Output\OutputInterface;
+
+/**
+ * Read-only inventory of a stack's storage backends, with a cleanup verdict per backend.
+ *
+ * Feeds manage:delete-backend — the ID lists printed at the end are ready to paste.
+ */
+class ListStorageBackends extends Command
```
[the remark this answers is not in the document]
Comment (jirkasemmler) [reply src/Keboola/Console/Command/ListStorageBackends.php:23]: Done in e9d0fca — row/verdict/filter/summary logika vytažena do plain `StorageBackendInventory`, command dělá jen I/O. Pokryto unit testy (tests/StorageBackendInventoryTest.php).

File: src/Keboola/Console/Command/ListStorageBackends.php (hunk 1 above)
Comment (jirkasemmler) [reply src/Keboola/Console/Command/ListStorageBackends.php]: `Client::getStorageBackend()` je v SDK až od v9.0.0 — tenhle repo pinuje `^7` (lock v7.1.1), kde detail metoda není a `apiGet` je private. Bump na ^10 jsem zkusil: composer projde, ale phpstan skočí z 8 na 236 errorů napříč ~20 dalšími commandy (v10 přidal nativní typy do signatur) — to je samostatná migrace mimo scope tohohle PR. Guzzle workaround je v e9d0fca aspoň zpevněný (přímá závislost, retry, timeouty); po bumpu klienta bude výměna za `getStorageBackend()` triviální.

File: src/Keboola/Console/Command/ListStorageBackends.php (hunk 2 above)
Comment (jirkasemmler) [reply src/Keboola/Console/Command/ListStorageBackends.php]: Opraveno v e9d0fca — summary i paste-ready ID listy se počítají z viditelných (filtrovaných) řádků; při aktivním filtru hlavička říká `Backends shown: X of Y (filters active — summary and ID lists cover the shown rows only)`. Regresní test `testSummarizeOnFilteredRowsCoversOnlyThoseRows`.

File: src/Keboola/Console/Command/ListStorageBackends.php (hunk 3 above)
Comment (jirkasemmler) [reply src/Keboola/Console/Command/ListStorageBackends.php:97]: Opraveno v e9d0fca — `--verdict` se validuje proti `StorageBackendInventory::VERDICTS`; neznámá hodnota = exit 1 s výpisem povolených verdiktů.

File: src/Keboola/Console/Command/ListStorageBackends.php (hunk 4 above)
Comment (jirkasemmler) [reply src/Keboola/Console/Command/ListStorageBackends.php]: Opraveno v e9d0fca — CSV se skládá přes `fputcsv` do php://temp a zapisuje se jedním `write(..., OUTPUT_RAW)`, console formatter se k obsahu buněk nedostane.

File: src/Keboola/Console/Command/ListStorageBackends.php (hunk 1 above)
Comment (jirkasemmler) [reply src/Keboola/Console/Command/ListStorageBackends.php]: Opraveno v e9d0fca — záznamy bez použitelného numerického id vrací `buildRow()` jako null, command je přeskočí a ohlásí na stderr (`skipped N list entries without a usable numeric id`).

File: src/Keboola/Console/Command/ListStorageBackends.php (hunk 5 above)
Comment (jirkasemmler) [reply src/Keboola/Console/Command/ListStorageBackends.php]: Opraveno v e9d0fca — `guzzlehttp/guzzle: ^7.0` přidán do require.

File: src/Keboola/Console/Command/ListStorageBackends.php (hunk 6 above)
Comment (jirkasemmler) [reply src/Keboola/Console/Command/ListStorageBackends.php]: e9d0fca přidává retry middleware (3 pokusy na 429/5xx/connect). Upstream metoda už existuje — `getStorageBackend()` od v9 — takže finální fix je bump klienta v samostatném PR (viz vlákno u Guzzle: bump na ^10 = 236 phpstan errorů napříč repem).

File: src/Keboola/Console/Command/ListStorageBackends.php (hunk 7 above)
Comment (jirkasemmler) [reply src/Keboola/Console/Command/ListStorageBackends.php]: Done v e9d0fca — plain `StorageBackendInventory` + `tests/StorageBackendInventoryTest.php`: verdiktová matice, filtry (vč. kombinace), buildRow edge cases, applyDetail, summarize vč. regresního testu, že filtrované delete listy nepouští skrytá ID. 24 testů, 86 asercí.

File: src/Keboola/Console/Command/ListStorageBackends.php (hunk 8 above)
Comment (jirkasemmler) [reply src/Keboola/Console/Command/ListStorageBackends.php]: Opraveno v e9d0fca — filtruje se nejdřív, detail se stahuje jen pro zobrazené řádky.

File: src/Keboola/Console/Command/ListStorageBackends.php (hunk 9 above)
Comment (jirkasemmler) [reply src/Keboola/Console/Command/ListStorageBackends.php:118]: Částečně adresováno v e9d0fca — detail se stahuje až po filtru, takže objem klesl na jednotky až desítky callů (stacky mají po úklidu ≤30 backendů). Sekvenční běh s retry jsem zatím nechal kvůli jednoduchosti; jestli na Poolu trváš, doplním s concurrency ~5.

File: src/Keboola/Console/Command/ListStorageBackends.php (hunk 10 above)
Comment (jirkasemmler) [reply src/Keboola/Console/Command/ListStorageBackends.php:209]: Opraveno v e9d0fca — hand-rolled escaping nahrazen `fputcsv` do php://temp (escape `''`, RFC 4180) + výpis přes OUTPUT_RAW.

File: src/Keboola/Console/Command/ListStorageBackends.php (hunk 1 above)
Comment (jirkasemmler) [reply src/Keboola/Console/Command/ListStorageBackends.php]: Update: klient bumpnut na ^10.3 v 50d8fec a Guzzle workaround kompletně nahrazen `Client::getStorageBackend()`. Typované signatury v10 odkryly 234 pre-existing phpstan nálezů ve 22 legacy commandech — ty jsou v `phpstan-baseline.neon` (level 9 dál platí pro nový kód), jejich úklid je mimo scope tohohle PR.

File: src/Keboola/Console/Command/ListStorageBackends.php (hunk 6 above)
Comment (jirkasemmler) [reply src/Keboola/Console/Command/ListStorageBackends.php]: Update: vyřešeno bumpem klienta v 50d8fec — detail jde přes `getStorageBackend()` s vestavěným exponenciálním backoffem klienta, custom retry middleware smazán.

File: phpstan-baseline.neon
Comment (romanbracinik) [inline phpstan-baseline.neon:1]: :eyes_bleeding:

**Review by romanbracinik — APPROVED** [review]
